File: services/enrichment/url-guard.ts
import '@/lib/server-guard';

import dns from 'node:dns/promises';
import net from 'node:net';

/**
 * SSRF guard for the enrichment crawler.
 *
 * Website URLs originate from third-party search providers, so they are
 * untrusted input. Before fetching, the hostname is resolved and every returned
 * address is checked against private, loopback, link-local and reserved ranges
 * so the crawler can never be aimed at internal infrastructure or cloud
 * metadata endpoints.
 */

function isPrivateIPv4(ip: string): boolean {
  const parts = ip.split('.').map(Number);
  if (parts.length !== 4 || parts.some((p) => !Number.isInteger(p) || p < 0 || p > 255)) {
    return true;
  }

  const [a, b] = parts;

  if (a === 0) return true; // "this network"
  if (a === 10) return true; // RFC1918
  if (a === 127) return true; // loopback
  if (a === 169 && b === 254) return true; // link-local, incl. cloud metadata
  if (a === 172 && b >= 16 && b <= 31) return true; // RFC1918
  if (a === 192 && b === 168) return true; // RFC1918
  if (a === 192 && b === 0) return true; // IETF protocol assignments
  if (a === 100 && b >= 64 && b <= 127) return true; // CGNAT
  if (a === 198 && (b === 18 || b === 19)) return true; // benchmarking
  if (a >= 224) return true; // multicast + reserved

  return false;
}

function isPrivateIPv6(ip: string): boolean {
  const address = ip.toLowerCase().split('%')[0];

  if (address === '::' || address === '::1') return true;
  if (address.startsWith('fe80') || address.startsWith('fec0')) return true; // link/site-local
  if (/^f[cd]/.test(address)) return true; // unique local
  if (address.startsWith('ff')) return true; // multicast

  // IPv4-mapped (::ffff:10.0.0.1) inherits the IPv4 rules.
  const mapped = address.match(/^::ffff:(\d+\.\d+\.\d+\.\d+)$/);
  if (mapped) return isPrivateIPv4(mapped[1]);

  return false;
}

export function isPrivateAddress(ip: string): boolean {
  const version = net.isIP(ip);
  if (version === 4) return isPrivateIPv4(ip);
  if (version === 6) return isPrivateIPv6(ip);
  return true;
}

const BLOCKED_HOST_SUFFIXES = ['.local', '.localhost', '.internal', '.home.arpa', '.onion'];

export interface CrawlTargetCheck {
  safe: boolean;
  /** Human-readable reason when `safe` is false, for the record's notes. */
  reason: string | null;
}

const SAFE: CrawlTargetCheck = { safe: true, reason: null };

/**
 * Resolves `hostname` and reports whether every address is publicly routable.
 *
 * Note this leaves a small TOCTOU window between the check and the fetch. It is
 * acceptable here because the crawler only ever reads public marketing pages and
 * discards the body, but a stricter deployment should pin the resolved address.
 */
export async function checkPubliclyRoutable(hostname: string): Promise<CrawlTargetCheck> {
  const host = hostname.toLowerCase().replace(/\.$/, '');

  if (!host || host === 'localhost') {
    return { safe: false, reason: 'Host is not publicly routable' };
  }
  if (BLOCKED_HOST_SUFFIXES.some((suffix) => host.endsWith(suffix))) {
    return { safe: false, reason: 'Host uses an internal-only domain suffix' };
  }

  // Literal IPs skip DNS entirely.
  if (net.isIP(host)) {
    return isPrivateAddress(host)
      ? { safe: false, reason: 'Host is a private or reserved IP address' }
      : SAFE;
  }

  let addresses: { address: string }[];
  try {
    addresses = await dns.lookup(host, { all: true, verbatim: true });
  } catch (error) {
    // A DNS failure is a dead or unreachable site, not a blocked one — say so,
    // because the two need very different follow-up.
    const code = (error as NodeJS.ErrnoException).code ?? 'DNS error';
    return { safe: false, reason: `Website domain could not be resolved (${code})` };
  }

  if (addresses.length === 0) {
    return { safe: false, reason: 'Website domain could not be resolved' };
  }
  if (addresses.some((entry) => isPrivateAddress(entry.address))) {
    return { safe: false, reason: 'Host resolves to a private or reserved IP address' };
  }

  return SAFE;
}

/** Validates a candidate crawl target: http(s) scheme on a public host. */
export async function checkCrawlTarget(url: string): Promise<CrawlTargetCheck> {
  let parsed: URL;
  try {
    parsed = new URL(url);
  } catch {
    return { safe: false, reason: 'Invalid website URL' };
  }

  if (parsed.protocol !== 'http:' && parsed.protocol !== 'https:') {
    return { safe: false, reason: 'Website URL is not HTTP or HTTPS' };
  }

  // Credentials in a URL are a redirect-abuse signal, not something a business site needs.
  if (parsed.username || parsed.password) {
    return { safe: false, reason: 'Website URL contains embedded credentials' };
  }

  return checkPubliclyRoutable(parsed.hostname);
}

File: backend/src/LeadMine.Infrastructure/Scraping/UrlGuard.cs
using System.Net;
using System.Net.Sockets;

namespace LeadMine.Infrastructure.Scraping;

public sealed record CrawlTargetCheck(bool Safe, string? Reason)
{
    public static readonly CrawlTargetCheck Allowed = new(true, null);
}

/// <summary>
/// SSRF guard for the enrichment crawler.
/// <para>
/// Website URLs come from third-party search providers, so they are untrusted
/// input. Before fetching, the hostname is resolved and every returned address
/// checked against private, loopback, link-local and reserved ranges — otherwise
/// a crafted listing could point this server at its own internal network or a
/// cloud metadata endpoint.
/// </para>
/// <para>
/// This matters more here than it did in the standalone worker: the crawler now
/// runs inside the API process, which sits on the same network as the database.
/// </para>
/// </summary>
public static class UrlGuard
{
    private static readonly string[] BlockedSuffixes =
        [".local", ".localhost", ".internal", ".home.arpa", ".onion"];

    public static async Task<CrawlTargetCheck> CheckAsync(string url, CancellationToken ct = default)
    {
        if (!Uri.TryCreate(url, UriKind.Absolute, out var uri))
        {
            return new CrawlTargetCheck(false, "Invalid website URL");
        }

        if (uri.Scheme != Uri.UriSchemeHttp && uri.Scheme != Uri.UriSchemeHttps)
        {
            return new CrawlTargetCheck(false, "Website URL is not HTTP or HTTPS");
        }

        // Credentials in a URL signal redirect abuse, not a business homepage.
        if (!string.IsNullOrEmpty(uri.UserInfo))
        {
            return new CrawlTargetCheck(false, "Website URL contains embedded credentials");
        }

        return await CheckHostAsync(uri.Host, ct);
    }

    private static async Task<CrawlTargetCheck> CheckHostAsync(string host, CancellationToken ct)
    {
        var name = host.ToLowerInvariant().TrimEnd('.');

        if (string.IsNullOrWhiteSpace(name) || name == "localhost")
        {
            return new CrawlTargetCheck(false, "Host is not publicly routable");
        }

        if (BlockedSuffixes.Any(suffix => name.EndsWith(suffix, StringComparison.Ordinal)))
        {
            return new CrawlTargetCheck(false, "Host uses an internal-only domain suffix");
        }

        // A literal IP skips DNS entirely.
        if (IPAddress.TryParse(name, out var literal))
        {
            return IsPrivate(literal)
                ? new CrawlTargetCheck(false, "Host is a private or reserved IP address")
                : CrawlTargetCheck.Allowed;
        }

        IPAddress[] addresses;
        try
        {
            addresses = await Dns.GetHostAddressesAsync(name, ct);
        }
        catch (SocketException ex)
        {
            // A dead domain is different from a blocked one, and the distinction
            // is what the operator needs to act on.
            return new CrawlTargetCheck(false, $"Website domain could not be resolved ({ex.SocketErrorCode})");
        }

        if (addresses.Length == 0)
        {
            return new CrawlTargetCheck(false, "Website domain could not be resolved");
        }

        return addresses.Any(IsPrivate)
            ? new CrawlTargetCheck(false, "Host resolves to a private or reserved IP address")
            : CrawlTargetCheck.Allowed;
    }

    public static bool IsPrivate(IPAddress address)
    {
        if (IPAddress.IsLoopback(address)) return true;

        if (address.AddressFamily == AddressFamily.InterNetwork)
        {
            var b = address.GetAddressBytes();

            return b[0] switch
            {
                0 => true,                                   // "this network"
                10 => true,                                  // RFC1918
                127 => true,                                 // loopback
                169 when b[1] == 254 => true,                // link-local + cloud metadata
                172 when b[1] >= 16 && b[1] <= 31 => true,   // RFC1918
                192 when b[1] == 168 => true,                // RFC1918
                192 when b[1] == 0 => true,                  // IETF assignments
                100 when b[1] >= 64 && b[1] <= 127 => true,  // CGNAT
                198 when b[1] is 18 or 19 => true,           // benchmarking
                >= 224 => true,                              // multicast + reserved
                _ => false,
            };
        }

        if (address.AddressFamily == AddressFamily.InterNetworkV6)
        {
            if (address.IsIPv6LinkLocal || address.IsIPv6SiteLocal || address.IsIPv6Multicast) return true;

            // IPv4-mapped addresses inherit the IPv4 rules.
            if (address.IsIPv4MappedToIPv6) return IsPrivate(address.MapToIPv4());

            var bytes = address.GetAddressBytes();

            // fc00::/7 unique local
            if ((bytes[0] & 0xFE) == 0xFC) return true;

            // ::
            if (bytes.All(b => b == 0)) return true;
        }

        return false;
    }
}
